syncthing: Fix systemd service

jokogr opened the pull request:
###### Motivation for this change

On some systems I tend to run syncthing only as a user, e.g.: 

```
  services.syncthing = {
    enable = true;
    systemService = false;
  };
```

I have noticed recently that the syncthing service was missing. It appears that due to a refactoring, the syncthing package has now multiple outputs (bin and out). This PR:

1. Fixes the executable paths in the systemd service files to the bin output.
2. Includes both outputs in `systemd.packages` (`bin` to have the executables in /nix/store and `out` to include properly the systemd files - the latter was the reason that the syncthing service was missing)
3. Stops creating service users if syncthing is not a system service.

I am still unsure whether to include syncthing.bin in `systemd.packages`, let me know if we can do it in another way, e.g. include it in `system.extraDependencies`.

Pinging @Mic92 since he refactored the syncthing package to use `buildGoPackage` in 643fabf8ea58f622943f6b90b43e9fb236bf35bc.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

jokogr committed:
syncthing: use proper output for binaries in services

jokogr committed:
syncthing: move systemd files to bin output

jokogr committed:
nixos/syncthing: setup user only on system service

Infinisil commented on nixos/modules/services/networking/syncthing.nix:
```diff
@@ -120,9 +120,9 @@ in {
     };
 
-    systemd.packages = [ pkgs.syncthing ];
+    systemd.packages = [ pkgs.syncthing.out pkgs.syncthing.bin ];
```
What's the problem if `pkgs.syncthing.bin` isn't included here?

jokogr replied:
Fair question, initially it was `pkgs.syncthing` which meant `pkgs.syncthing.bin`. This was a problem, since the `bin` output does not have the systemd service files.

By using  `pkgs.syncthing.out` the reference to the binaries got lost, so I thought to include `bin` as a systemd.package.

Infinisil replied:
Looked it up, `systemd.packages` (as I thought) only uses `$out/etc/systemd` of the packages, it only adds systemd units defined in them:

https://github.com/NixOS/nixpkgs/blob/830599debecac99f0215029c922879af84be386e/nixos/modules/system/boot/systemd-lib.nix#L150-L152

So there won't be any problems with removing that (and `pkgs.syncthing.out = pkgs.syncthing`)

jokogr replied:
> So there won't be any problems with removing that (and `pkgs.syncthing.out = pkgs.syncthing`)

I am afraid this is not correct, from my tests it is more like `pkgs.syncthing.bin = pkgs.syncthing`. 

My impression is that `buildGoPackage` in `pkgs/development/go-modules/generic/default.nix` sets

```
 outputs = args.outputs or [ "bin" "out" ];
```

so the first output ("bin") is picked when output is not defined (a.k.a. `pkgs.syncthing`).

So, we have to explicitly declare the "out" output in the `systemd.packages` and then again we are missing a reference to the other output, which contains binaries.

Infinisil replied:
Ohh, I didn't check, you're right. `pkgs.syncthing.out` in `systemd.packages` should be good then. The binaries aren't needed in `systemd.packages` though, so that can be removed.

jokogr replied:
ok, good thing I understood `out` well.

Now, I agree that binaries are not needed in `systemd.packages`, but if I leave `syncthing.bin` out, then it is not referred anywhere else, so it is not installed. This is why I kept it.

Other approaches would be:

1. To include bin in `environment.systemPackages`.
2. To include bin in `system.extraDependencies` as mentioned in motivation.
3. To move the systemd files to $bin as suggested by @Mic92.

Personally I find 2 the most elegant, but it is mentioned that this option is primarily used by the installation tests. In the end, I will follow @Mic92's suggestion.

Infinisil replied:
In what way does putting it in `systemd.packages` make it get installed? It shouldn't. If by "install" you mean having it accessible in `$PATH`, then `environment.systemPackages` is what you want (and I suspect you do). In NixOS it doesn't make sense to "install" something just to have it be built. Either it's needed (and then you refer to it where you need it and it gets available like this), or it's not needed (so you don't refer to it and it won't be available). There's no notion of installing it just to have it installed.

Mic92 commented:
Can you move all outputs to `$bin` instead?

Mic92 commented:
@GrahamcOfBorg build syncthing

rnhmjoj commented:
status?

jokogr commented:
* Followed @Mic92's piece of advice and moved the systemd unit files to the `bin` output.
* Rebased to current master.
* Retested the service on my machine, it still works as intended.

Infinisil commented:
Oh, can you format your last commit message with the `nixos/syncthing:` prefix?

ikoutras-epages commented:
Amended and rebased once again :)

jokogr commented:
Amended and rebased once again :)
